src/backend/vulkan/src/window.rs

```diff
@@ -16,13 +16,102 @@ use winit;
+            Connection::Xlib(dpy, window) => unsafe {
+                use x11::xlib::{XGetWindowAttributes, XWindowAttributes};
+                use std::mem::zeroed;
+                let mut attribs: XWindowAttributes = zeroed();
+                let result = XGetWindowAttributes(*dpy as _, *window, &mut attribs);
+                if result == 0 {
+                    panic!("XGetGeometry failed");
+                }
+                hal::window::Extent2D {
+                    width: attribs.width as Size,
+                    height: attribs.height as Size,
+                }
+            },
+            #[cfg(all(unix, not(target_os = "android")))]
+            Connection::Xcb(connection, window) => unsafe {
+                use std::mem;
+                use xcb::{Connection, xproto};
+                let conn = Connection::from_raw_conn(*connection as _);
+                let geometry = xproto::get_geometry(&conn, *window)
+                    .get_reply()
+                    .expect("xcb_get_geometry failed")
+                    .ptr
+                    .as_ref()
+                    .expect("unexpected NULL XCB geometry");
+                mem::forget(conn); //TODO: use `into_raw_conn`
+                hal::window::Extent2D {
+                    width: geometry.width as Size,
+                    height: geometry.height as Size,
+                }
+            },
+            #[cfg(all(unix, not(target_os = "android")))]
+            Connection::Wayland(window) => {
+                let (width, height) = window.get_extent().unwrap();
+                hal::window::Extent2D { width, height }
+            },
+            #[cfg(target_os = "android")]
+            Connection::Android(window) => {
+                use winit::os::android::WindowExt;
+                let (width, height) = window.get_extent().unwrap();
+                hal::window::Extent2D { width, height }
+            },
+            #[cfg(windows)]
+            Connection::Hwnd(hwnd) => unsafe {
+                use winapi::shared::windef::RECT;
+                use winapi::um::winuser::GetClientRect;
+                use std::mem::zeroed;
+
+                let mut rect: RECT = zeroed();
+                if GetClientRect(*hwnd as *mut _, &mut rect as *mut RECT) == 0 {
+                    panic!("GetClientRect failed");
+                }
+                hal::window::Extent2D {
+                    width: (rect.right - rect.left) as Size,
+                    height: (rect.bottom - rect.top) as Size,
+                }
+            },
+        }
+    }
+}
 
 pub struct Surface {
     // Vk (EXT) specs [29.2.7 Platform-Independent Information]
     // For vkDestroySurfaceKHR: Host access to surface must be externally synchronized
     pub(crate) raw: Arc<RawSurface>,
-    pub(crate) width: Size,
-    pub(crate) height: Size,
+    pub(crate) current_extent: Option<hal::window::Extent2D>,
+    connection: Arc<Mutex<Connection>>,
```

Comment:
who is this shared with?

Reply:
Nobody as far as I know, the mutex is only used by me insure Sync & Send get implemented. Or do you mean the Arc? That's so that we don't have to use lifetimes which can't be used due to lack of type ascription.

Reply:
I mean, could this just be `connection: Connection`?